ssl/quic/quic_txp.c

```diff
@@ -0,0 +1,2027 @@
+{
+    OSSL_QUIC_TX_PACKETISER *txp;
+
+    if (args == NULL
+        || args->qtx == NULL
+        || args->txpim == NULL
+        || args->cfq == NULL
+        || args->ackm == NULL
+        || args->qsm == NULL
+        || args->conn_txfc == NULL
+        || args->conn_rxfc == NULL)
+        return 0;
+
+    txp = OPENSSL_zalloc(sizeof(*txp));
+    if (txp == NULL)
+        return NULL;
+
+    txp->args           = *args;
+    txp->last_tx_time   = ossl_time_zero();
+
+    if (!ossl_quic_fifd_init(&txp->fifd,
+                             txp->args.cfq, txp->args.ackm, txp->args.txpim,
+                             get_sstream_by_id, txp,
+                             on_regen_notify, txp)) {
+        OPENSSL_free(txp);
+        return NULL;
+    }
+
+    return txp;
+}
+
+void ossl_quic_tx_packetiser_free(OSSL_QUIC_TX_PACKETISER *txp)
+{
+    if (txp == NULL)
+        return;
+
+    ossl_quic_tx_packetiser_set_initial_token(txp, NULL, 0, NULL, NULL);
+    ossl_quic_fifd_cleanup(&txp->fifd);
+    OPENSSL_free(txp->iovec);
+    OPENSSL_free((char *)txp->conn_close_frame.reason);
+    OPENSSL_free(txp->scratch);
+    OPENSSL_free(txp);
+}
+
+void ossl_quic_tx_packetiser_set_initial_token(OSSL_QUIC_TX_PACKETISER *txp,
+                                               const unsigned char *token,
+                                               size_t token_len,
+                                               ossl_quic_initial_token_free_fn *free_cb,
+                                               void *free_cb_arg)
+{
+    if (txp->initial_token != NULL && txp->initial_token_free_cb != NULL)
+        txp->initial_token_free_cb(txp->initial_token, txp->initial_token_len,
+                                   txp->initial_token_free_cb_arg);
+
+    txp->initial_token              = token;
+    txp->initial_token_len          = token_len;
+    txp->initial_token_free_cb      = free_cb;
+    txp->initial_token_free_cb_arg  = free_cb_arg;
+}
+
+int ossl_quic_tx_packetiser_set_cur_dcid(OSSL_QUIC_TX_PACKETISER *txp,
+                                         const QUIC_CONN_ID *dcid)
+{
+    if (dcid == NULL)
+        return 0;
+
+    txp->args.cur_dcid = *dcid;
+    return 1;
+}
+
+int ossl_quic_tx_packetiser_set_cur_scid(OSSL_QUIC_TX_PACKETISER *txp,
+                                         const QUIC_CONN_ID *scid)
+{
+    if (scid == NULL)
+        return 0;
+
+    txp->args.cur_scid = *scid;
+    return 1;
+}
+
+/* Change the destination L4 address the TXP uses to send datagrams. */
+int ossl_quic_tx_packetiser_set_peer(OSSL_QUIC_TX_PACKETISER *txp,
+                                     const BIO_ADDR *peer)
+{
+    if (peer == NULL)
+        return 0;
+
+    txp->args.peer = *peer;
+    return 1;
+}
+
+int ossl_quic_tx_packetiser_discard_enc_level(OSSL_QUIC_TX_PACKETISER *txp,
+                                              uint32_t enc_level)
+{
+    if (enc_level >= QUIC_ENC_LEVEL_NUM)
+        return 0;
```

Comment:
Probably a general comment for a lot of our QUIC code. Shouldn't we be calling `ERR_raise` for all of these errors?